lib/std/symbol_map_unix.zig

```
@@ -0,0 +1,172 @@
+                        // main program
+                        context.name = mem.spanZ(info.dlpi_name) orelse "";
+                        context.base_address = info.dlpi_addr;
+                        // Stop the iteration
+                        return error.Found;
+                    }
+                }
+            }
+        }.callback)) {
+            return BaseError.MissingDebugInfo;
+        } else |err| switch (err) {
+            error.Found => {},
+            else => return BaseError.MissingDebugInfo,
+        }
+
+        if (address_map.get(ctx.base_address)) |obj_di| {
+            return obj_di;
+        }
+
+        const obj_di = try allocator.create(Self);
+        errdefer allocator.destroy(obj_di);
+
+        // TODO https://github.com/ziglang/zig/issues/5525
+        const copy = if (ctx.name.len > 0)
+            fs.cwd().openFile(ctx.name, .{ .intended_io_mode = .blocking })
+        else
+            fs.openSelfExe(.{ .intended_io_mode = .blocking });
+
+        const elf_file = copy catch |err| switch (err) {
+            error.FileNotFound => return BaseError.MissingDebugInfo,
+            else => return err,
+        };
+
+        obj_di.* = try readElfDebugInfo(allocator, elf_file);
+        obj_di.base_address = ctx.base_address;
+
+        try address_map.putNoClobber(ctx.base_address, obj_di);
+
+        return obj_di;
+    }
+
+    /// This takes ownership of elf_file: users of this function should not close
+    /// it themselves, even on error.
+    /// TODO resources https://github.com/ziglang/zig/issues/4353
+    /// TODO it's weird to take ownership even on error, rework this code.
+    fn readElfDebugInfo(allocator: *mem.Allocator, elf_file: File) !Self {
+        nosuspend {
+            const mapped_mem = try debug_info.mapWholeFile(elf_file);
+            const hdr = @ptrCast(*const elf.Ehdr, &mapped_mem[0]);
+            if (!mem.eql(u8, hdr.e_ident[0..4], "\x7fELF")) return error.InvalidElfMagic;
```

While looking into how to properly turn things into interfaces, I noticed that a lot of the error handling is a bit odd.
It seems like most of the errors which could be bubbled up via `addressToSymbol` are really just symptoms of the debug symbols being invalid. For instance, this `InvalidElfMagic` error would be bubbled up and printed out halting printing the rest of the stack trace. This will only  only occur if `elf_file` is invalid.

So how do we want to handle errors like this? Silently swallow? Bubble them up and stop printing? Note that this function is private and only used for `SymbolMap`.

---

As an aside, https://github.com/ziglang/zig/issues/2647 would be useful for making the interface simple if wanted to support many "types" of errors.

---

At the moment, I have just used `anyerror`. This should (probably?) be changed before merge.